dart/gui/osg/RealTimeWorldNode.cpp

```diff
@@ -0,0 +1,166 @@
+void RealTimeWorldNode::setTargetFrequency(double targetFrequency)
+{
+  const double targetRTF = mTargetSimTimeLapse/mTargetRealTimeLapse;
+  mTargetRealTimeLapse = 1.0/targetFrequency;
+  mTargetSimTimeLapse = targetRTF*mTargetRealTimeLapse;
+}
+
+//==============================================================================
+double RealTimeWorldNode::getTargetFrequency() const
+{
+  return 1.0/mTargetRealTimeLapse;
+}
+
+//==============================================================================
+void RealTimeWorldNode::setTargetRealTimeFactor(double targetRTF)
+{
+  mTargetSimTimeLapse = targetRTF*mTargetRealTimeLapse;
+}
+
+//==============================================================================
+double RealTimeWorldNode::getTargetRealTimeFactor() const
+{
+  return mTargetSimTimeLapse/mTargetRealTimeLapse;
+}
+
+//==============================================================================
+double RealTimeWorldNode::getLastRealTimeFactor() const
+{
+  return mLastRealTimeFactor;
+}
+
+//==============================================================================
+double RealTimeWorldNode::getLowestRealTimeFactor() const
+{
+  return mLowestRealTimeFactor;
+}
+
+//==============================================================================
+double RealTimeWorldNode::getHighestRealTimeFactor() const
+{
+  return mHighestRealTimeFactor;
+}
+
+//==============================================================================
+void RealTimeWorldNode::refresh()
+{
+  customPreRefresh();
+  clearChildUtilizationFlags();
+
+  if (mNumStepsPerCycle != 1)
+  {
+    dtwarn << "[RealTimeWorldNode] The number of steps per cycle has been set "
+           << "to [" << mNumStepsPerCycle << "], but this value is ignored by "
+           << "the RealTimeWorldNode::refresh() function. Use the function "
+           << "RealTimeWorldNode::setTargetRealTimeFactor(double) to change "
+           << "the simulation speed\n";
+    mNumStepsPerCycle = 1;
+  }
+
+  if (mFirstRefresh)
+  {
+    mRefreshTimer.setStartTick();
+    mFirstRefresh = false;
+  }
+
+  if (mWorld && mSimulating)
+  {
+    const double startSimTime = mWorld->getTime();
+    const double simTimeStep = mWorld->getTimeStep();
+
+    while(mRefreshTimer.time_s() < mTargetRealTimeLapse)
+    {
+      const double nextSimTimeLapse =
+          mWorld->getTime() - startSimTime + simTimeStep;
+
+      if(nextSimTimeLapse <= mTargetSimTimeLapse)
+      {
+        customPreStep();
+        mWorld->step();
+        customPostStep();
+      }
+    }
+
+    mLastRealTimeFactor =
+        (mWorld->getTime() - startSimTime)/mTargetRealTimeLapse;
+    mLowestRealTimeFactor =
+        std::min(mLastRealTimeFactor, mLowestRealTimeFactor);
+    mHighestRealTimeFactor =
+        std::max(mLastRealTimeFactor, mHighestRealTimeFactor);
+  }
+  else
+  {
+    mFirstRefresh = true;
+  }
+
+  mRefreshTimer.setStartTick();
```

Comment:
Please correct me if I'm wrong:

This might be micro optimization, but we could reduce the range of use of `mRefreshTimer` by putting `if (mFirstRefresh) {...}` and this line (`mRefreshTimer.setStartTick();`) into `if (mWorld && mSimulating) {...}` like:

```cpp
  if (mWorld && mSimulating)
  {
    if (mFirstRefresh)
    {
      mRefreshTimer.setStartTick();
      mFirstRefresh = false;
    }

    const double startSimTime = mWorld->getTime();
    const double simTimeStep = mWorld->getTimeStep();

    while(mRefreshTimer.time_s() < mTargetRealTimeLapse)
    {
      const double nextSimTimeLapse =
          mWorld->getTime() - startSimTime + simTimeStep;

      if(nextSimTimeLapse <= mTargetSimTimeLapse)
      {
        customPreStep();
        mWorld->step();
        customPostStep();
      }
    }

    mLastRealTimeFactor =
        (mWorld->getTime() - startSimTime)/mTargetRealTimeLapse;
    mLowestRealTimeFactor =
        std::min(mLastRealTimeFactor, mLowestRealTimeFactor);
    mHighestRealTimeFactor =
        std::max(mLastRealTimeFactor, mHighestRealTimeFactor);

    mRefreshTimer.setStartTick();
  }
  else
  {
    mFirstRefresh = true;
  }
```

By this change, we could save calling `mRefreshTimer.setStartTick();` twice when it's not in simulating. Not sure how heavy to call `mRefreshTimer.setStartTick()` though.

Reply:
It's a bit of a micro-optimization, but there's nothing wrong with that. If it seems more logical to you to have them inside the if-statement, I have no problem with that :+1: